File: car_diary.rb
# car_diary.rb — Ruby версия

require 'json'
require 'date'

DATA_FILE = 'repairs.json'

class CarDiary
  attr_reader :repairs

  def initialize
    @repairs = []
    load_data
  end

  def load_data
    if File.exist?(DATA_FILE)
      begin
        @repairs = JSON.parse(File.read(DATA_FILE), symbolize_names: true)
      rescue
        @repairs = []
      end
    else
      @repairs = []
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@repairs))
  end

  def add(date, mileage, desc, cost)
    repair = {
      id: @repairs.size + 1,
      date: date,
      mileage: mileage,
      description: desc,
      cost: cost
    }
    @repairs << repair
    save
    repair[:id]
  end

  def list
    if @repairs.empty?
      puts "\e[33mНет записей.\e[0m"
      return
    end
    puts "\e[36m%-4s %-12s %-10s %-30s %10s\e[0m" % ["ID", "Дата", "Пробег", "Описание", "Стоимость"]
    puts "-" * 70
    @repairs.each do |r|
      puts "%-4d %-12s %-10d %-30s %10.2f" % [r[:id], r[:date], r[:mileage], r[:description], r[:cost]]
    end
  end

  def search(keyword)
    results = @repairs.select do |r|
      r[:description].downcase.include?(keyword.downcase) ||
      r[:date].include?(keyword) ||
      r[:mileage].to_s == keyword
    end
    if results.empty?
      puts "\e[33mНичего не найдено.\e[0m"
    else
      results.each { |r| puts "#{r[:id]}: #{r[:date]} | #{r[:mileage]} км | #{r[:description]} | #{r[:cost]} руб." }
    end
  end

  def delete(id)
    found = @repairs.find { |r| r[:id] == id }
    if found
      @repairs.delete(found)
      save
      true
    else
      false
    end
  end

  def stats
    if @repairs.empty?
      puts "Нет данных для статистики."
      return
    end
    total = @repairs.sum { |r| r[:cost] }
    puts "\e[32mОбщая стоимость ремонтов: #{total} руб.\e[0m"
    avg = total / @repairs.size
    puts "Средняя стоимость ремонта: #{avg.round(2)} руб."
    if @repairs.size > 1
      mileages = @repairs.map { |r| r[:mileage] }
      intervals = (1...mileages.size).map { |i| mileages[i] - mileages[i-1] }
      avg_interval = intervals.sum / intervals.size.to_f
      puts "Средний интервал между ремонтами: #{avg_interval.round(1)} км"
    end
  end
end

def main
  diary = CarDiary.new
  loop do
    puts "\n\e[36m🚗 Дневник автомобиля (Ruby)\e[0m"
    puts "1. Добавить запись"
    puts "2. Показать все записи"
    puts "3. Поиск записей"
    puts "4. Удалить запись"
    puts "5. Статистика"
    puts "6. Выход"
    print "Выберите действие: "
    choice = gets.chomp
    case choice
    when "1"
      print "Дата (ГГГГ-ММ-ДД): "
      date = gets.chomp
      print "Пробег (км): "
      mileage = gets.chomp.to_i
      print "Описание: "
      desc = gets.chomp
      print "Стоимость (руб): "
      cost = gets.chomp.to_f
      id = diary.add(date, mileage, desc, cost)
      puts "\e[32m✅ Запись добавлена (ID: #{id})\e[0m"
    when "2"
      diary.list
    when "3"
      print "Введите ключевое слово: "
      keyword = gets.chomp
      diary.search(keyword)
    when "4"
      diary.list
      print "Введите ID записи для удаления: "
      id = gets.chomp.to_i
      if diary.delete(id)
        puts "\e[32m✅ Запись удалена.\e[0m"
      else
        puts "\e[31m❌ Запись не найдена.\e[0m"
      end
    when "5"
      diary.stats
    when "6"
      puts "До свидания!"
      break
    else
      puts "\e[31mНеверный выбор.\e[0m"
    end
  end
end

main if __FILE__ == $0

File: car_diary.py
# car_diary.py — Python версия

import json
import os
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)
DATA_FILE = "repairs.json"

class CarDiary:
    def __init__(self):
        self.repairs = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r', encoding='utf-8') as f:
                    self.repairs = json.load(f)
            except:
                self.repairs = []

    def save(self):
        with open(DATA_FILE, 'w', encoding='utf-8') as f:
            json.dump(self.repairs, f, indent=2, ensure_ascii=False)

    def add(self, date, mileage, desc, cost):
        repair = {
            "id": len(self.repairs) + 1,
            "date": date,
            "mileage": mileage,
            "description": desc,
            "cost": cost
        }
        self.repairs.append(repair)
        self.save()
        return repair["id"]

    def list_all(self):
        if not self.repairs:
            print(Fore.YELLOW + "Нет записей.")
            return
        print(Fore.CYAN + f"{'ID':<4} {'Дата':<12} {'Пробег':<10} {'Описание':<30} {'Стоимость':>10}")
        print("-" * 70)
        for r in self.repairs:
            print(f"{r['id']:<4} {r['date']:<12} {r['mileage']:<10} {r['description']:<30} {r['cost']:>10}")

    def search(self, keyword):
        results = [r for r in self.repairs if keyword.lower() in r['description'].lower() or keyword in r['date'] or str(r['mileage']) == keyword]
        if not results:
            print(Fore.YELLOW + "Ничего не найдено.")
        else:
            for r in results:
                print(f"{r['id']}: {r['date']} | {r['mileage']} км | {r['description']} | {r['cost']} руб.")

    def delete(self, id):
        for i, r in enumerate(self.repairs):
            if r['id'] == id:
                del self.repairs[i]
                self.save()
                return True
        return False

    def stats(self):
        total = sum(r['cost'] for r in self.repairs)
        print(Fore.GREEN + f"Общая стоимость ремонтов: {total} руб.")
        if self.repairs:
            avg = total / len(self.repairs)
            print(f"Средняя стоимость ремонта: {avg:.2f} руб.")
            # пробеги
            mileages = [r['mileage'] for r in self.repairs]
            if len(mileages) > 1:
                intervals = [mileages[i] - mileages[i-1] for i in range(1, len(mileages))]
                avg_interval = sum(intervals) / len(intervals)
                print(f"Средний интервал между ремонтами: {avg_interval:.1f} км")
        else:
            print("Нет данных для статистики.")

def main():
    diary = CarDiary()
    while True:
        print(Fore.CYAN + "\n🚗 Дневник автомобиля (Python)")
        print("1. Добавить запись")
        print("2. Показать все записи")
        print("3. Поиск записей")
        print("4. Удалить запись")
        print("5. Статистика")
        print("6. Выход")
        choice = input("Выберите действие: ").strip()
        if choice == "1":
            date = input("Дата (ГГГГ-ММ-ДД): ")
            mileage = int(input("Пробег (км): "))
            desc = input("Описание: ")
            cost = float(input("Стоимость (руб): "))
            id = diary.add(date, mileage, desc, cost)
            print(Fore.GREEN + f"✅ Запись добавлена (ID: {id})")
        elif choice == "2":
            diary.list_all()
        elif choice == "3":
            keyword = input("Введите ключевое слово (дата, пробег, описание): ")
            diary.search(keyword)
        elif choice == "4":
            diary.list_all()
            id = int(input("Введите ID записи для удаления: "))
            if diary.delete(id):
                print(Fore.GREEN + "✅ Запись удалена.")
            else:
                print(Fore.RED + "❌ Запись не найдена.")
        elif choice == "5":
            diary.stats()
        elif choice == "6":
            print("До свидания!")
            break
        else:
            print(Fore.RED + "Неверный выбор, попробуйте снова.")

if __name__ == "__main__":
    main()
